discourse_check_cleanup.py

```diff
@@ -0,0 +1,306 @@
+            expected_count,
+            url_count,
+            urls_with_actions,
+        )
+        return False
+    return True
+
+
+def _check_url_retrieve(
+    urls_with_actions: dict[str, str], discourse: Discourse, test_name: str
+) -> bool:
+    """Check that retrieving the URL succeeds.
+
+    Args:
+        urls_with_actions: The URLs that had any actions against them.
+        discourse: Client to the documentation server.
+        test_name: The name of the test to include in the logging message.
+
+    Returns:
+        Whether the test succeeded.
+    """
+    for url in urls_with_actions.keys():
+        try:
+            discourse.retrieve_topic(url=url)
+        except DiscourseError as exc:
+            logging.error(
+                "%s check failed, URL retrieval failed for %s, error: %s, urls_with_actions=%s",
+                test_name,
+                url,
+                exc,
+                urls_with_actions,
+            )
+            return False
+    return True
+
+
+def _check_url_result(
+    urls_with_actions: dict[str, str], expected_result: list[str], test_name: str
+) -> bool:
+    """Check the results for the URLs.
+
+    Args:
+        urls_with_actions: The URLs that had any actions against them.
+        expected_result: The expected results.
+        test_name: The name of the test to include in the logging message.
+
+    Returns:
+        Whether the test succeeded.
+    """
+    if sorted(results := urls_with_actions.values()) != sorted(expected_result):
+        logging.error(
+            "%s check failed, the result is not as expected, "
+            "expected: %s, got: %s, urls_with_actions=%s",
+            test_name,
+            results,
+            expected_result,
+            urls_with_actions,
+        )
+        return False
+    return True
+
+
+def check_draft(urls_with_actions: dict[str, str]) -> bool:
+    """Check that the draft test succeeded.
+
+    Success is indicated by that there are no URLs with actions.
+
+    Args:
+        urls_with_actions: The URLs that had any actions against them.
+
+    Returns:
+        Whether the test succeeded.
+    """
+    test_name = "draft"
+    if not _check_url_count(
+        urls_with_actions=urls_with_actions, expected_count=0, test_name=test_name
+    ):
+        return False
+
+    logging.info("%s check succeeded", test_name)
+    return True
+
+
+def check_create(urls_with_actions: dict[str, str], discourse: Discourse) -> bool:
+    """Check that the create test succeeded.
+
+    Success is indicated by that there are 3 URLs with a success result and that retrieving the
+    URLs succeeds.
+
+    Args:
+        urls_with_actions: The URLs that had any actions against them.
+        discourse: Client to the documentation server.
+
+    Returns:
+        Whether the test succeeded.
+    """
+    test_name = "create"
+    if not _check_url_count(
+        urls_with_actions=urls_with_actions, expected_count=3, test_name=test_name
```

Comment:
The trade-off here is complexity. There are 2 kinds of integration tests in this repository, there are those defined in `pytest` and the integration tests defined in the workflow, for which this is a utility. For clarity, let's call these end to end tests since they are intended to simulate how a user would run this action end to end.

The main reasons for the end to end tests to exist are (1) to check that the parameters are correctly being passed from the action input to the code and (2) to check the interactions against the charmhub discourse server rather than a server spun up just for testing. This means that the end to end tests are likely to be fairly stable. Parameters would be useful if these tests were expected to change a lot or we would need to run many different similar tests that just vary by the parameters we add. In this case, we would need to add optional parameters for the input to the tests and then verify whether they have been passed in correctly etc versus simply treating the functions in this utility as the `assert` block for normal tests and the `integration_test.yaml` as the `arrange` and `act` block. It would be great if this utility wasn't needed, although coding these checks in bash would seem to be more complex which is why this utility was created.

Any feedback on this?

Reply:
I really get the complexity of generalizing these checks. The way I though about it was passing an array as an extra argument or extend the "urls_with_actions" to add the expected result for this url for this test
And then add the logic to count the number of url in the argument and compare the result with the expected result

The shouldn't add too much complexity and allows the test to be more generic

Reply:
The tests now take an additional parameter `--action-kwargs` which take the expected results for each of the actions which has made them more generic. At this point it does not include a parameter to indicate which URLs should and should not be able to be retrieved which would add a lot more complexity, In all but 1 test it is all URLs should be retrievable anyway

Reply:
Marking as resolved for now unless there is any further feedback?